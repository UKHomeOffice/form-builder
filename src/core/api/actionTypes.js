const prefix = 'useApiRequest/';

export const EXECUTING = `${prefix}EXECUTING`;
export const SUCCESS = `${prefix}SUCCESS`;
export const ERROR = `${prefix}ERROR`;
