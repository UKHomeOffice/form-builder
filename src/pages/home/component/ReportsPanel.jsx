import {Segment} from "semantic-ui-react";
import {Bar, BarChart, CartesianGrid, Legend, Pie, PieChart, Sector, Tooltip, XAxis, YAxis} from "recharts";
import React, {useState} from "react";

const ReportsPanel = () => {
    const [activeIndex, setActiveIndex] = useState(0);

    const onPieEnter = (data, index) => {
        setActiveIndex(index);
    };
    const data = [
        {name: 'dev', value: 400},
        {name: 'demo', value: 300},
        {name: 'staging', value: 300},
        {name: 'prod', value: 200},
    ];


    const formTypes = [
        {
            name: 'dev', wizard: 10, form: 10
        },
        {
            name: 'demo', wizard: 110, form: 10
        },
        {
            name: 'staging', wizard: 140, form: 10
        },
        {
            name: 'prod', wizard: 10, form: 100
        },

    ];

    const renderActiveShape = (props) => {
        const RADIAN = Math.PI / 180;
        const {
            cx, cy, midAngle, innerRadius, outerRadius, startAngle, endAngle,
            fill, payload, value,
        } = props;
        const sin = Math.sin(-RADIAN * midAngle);
        const cos = Math.cos(-RADIAN * midAngle);
        const sx = cx + (outerRadius + 10) * cos;
        const sy = cy + (outerRadius + 10) * sin;
        const mx = cx + (outerRadius + 30) * cos;
        const my = cy + (outerRadius + 30) * sin;
        const ex = mx + (cos >= 0 ? 1 : -1) * 22;
        const ey = my;
        const textAnchor = cos >= 0 ? 'start' : 'end';
        return (
            <g>
                <text x={cx} y={cy} dy={8} textAnchor="middle" fill={fill}>{payload.name}</text>
                <Sector
                    cx={cx}
                    cy={cy}
                    innerRadius={innerRadius}
                    outerRadius={outerRadius}
                    startAngle={startAngle}
                    endAngle={endAngle}
                    fill={fill}
                />
                <Sector
                    cx={cx}
                    cy={cy}
                    startAngle={startAngle}
                    endAngle={endAngle}
                    innerRadius={outerRadius + 6}
                    outerRadius={outerRadius + 10}
                    fill={fill}
                />
                <path d={`M${sx},${sy}L${mx},${my}L${ex},${ey}`} stroke={fill} fill="none"/>
                <circle cx={ex} cy={ey} r={2} fill={fill} stroke="none"/>
                <text x={ex + (cos >= 0 ? 1 : -1) * 12} y={ey} textAnchor={textAnchor}
                      fill="#333">{`${value} forms`}</text>
            </g>
        );
    };

    return  <Segment.Group horizontal basic>
        <Segment basic>
            <PieChart width={500} height={400}>
                <Pie
                    activeIndex={activeIndex}
                    activeShape={renderActiveShape}
                    data={data}
                    cx={200}
                    cy={200}
                    innerRadius={60}
                    outerRadius={80}
                    fill="#8884d8"
                    dataKey="value"
                    onMouseEnter={onPieEnter}
                />
            </PieChart>
        </Segment>
        <Segment basic>
            <BarChart
                width={500}
                height={400}
                data={formTypes}
                margin={{
                    top: 20, right: 30, left: 20, bottom: 5,
                }}>
                <CartesianGrid strokeDasharray="3 3"/>
                <XAxis dataKey="name"/>
                <YAxis/>
                <Tooltip/>
                <Legend/>
                <Bar dataKey="wizard" stackId="a" fill="#8884d8"/>
                <Bar dataKey="form" stackId="a" fill="#82ca9d"/>
            </BarChart>
        </Segment>
    </Segment.Group>
}

export default ReportsPanel;
