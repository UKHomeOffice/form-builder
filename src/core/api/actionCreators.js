import { EXECUTING, SUCCESS, ERROR } from './actionTypes';

export const executing = () => ({ type: EXECUTING });
export const success = response => ({ type: SUCCESS, response });
export const error = (response, exception) => ({
    type: ERROR,
    response,
    exception,
});
