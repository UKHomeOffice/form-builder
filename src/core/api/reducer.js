import { EXECUTING, SUCCESS, ERROR } from './actionTypes';

export const initialState = {
    status: null,
    response: null,
    exception: null,
};

const reducer = (state = initialState, { type, response, exception } = {}) => {
    switch (type) {
        case EXECUTING:
            return { ...initialState, status: EXECUTING };
        case SUCCESS:
            return { ...state, status: SUCCESS, response };
        case ERROR:
            return {
                ...state,
                status: ERROR,
                response,
                exception,
            };
        default:
            return state;
    }
};

export default reducer;
