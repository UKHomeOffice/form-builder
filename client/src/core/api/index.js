import {useReducer} from 'react';
import axios from 'axios';
import reducer, {initialState} from './reducer';
import {error, executing, success} from './actionCreators';
import useEnvContext from "../context/useEnvContext";
import {useKeycloak} from "react-keycloak";
import keycloakTokenProvider from '../auth/KeycloakTokenProvider';
import qs from "querystring";
import appConfig from 'react-global-configuration';


const configureAxios = async (envContext, config, keycloak) => {
    config.headers['Accept'] = 'application/json';
    config.headers['Content-Type'] = 'application/json';
    config.headers['Cache-Control'] = "no-cache";
    config.headers['x-user-email'] = keycloak.tokenParsed.email;
    const jwtToken = await keycloakTokenProvider.fetchKeycloakToken(envContext, keycloak);
    config.headers['Authorization'] = `Bearer ${jwtToken}`;
    return Promise.resolve(config);
};

export const useProxyApiRequest = (path, {verb = 'get', params = {}} = {}) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const instance = axios.create();
    const [keycloak] = useKeycloak();
    const keycloakConfig = appConfig.get('keycloak');

    instance.interceptors.request.use(async (config) => {
            config.headers['x-user-email'] = keycloak.tokenParsed.email;
            const accessToken = await axios({
                method: 'POST',
                url: `${keycloakConfig.authUrl}/realms/${keycloakConfig.realm}/protocol/openid-connect/token`,
                headers: {
                    "Content-Type": "application/x-www-form-urlencoded"
                },
                data: qs.stringify({
                    grant_type: 'refresh_token',
                    client_id: keycloakConfig.clientId,
                    refresh_token: keycloak.refreshToken
                })
            });
            config.headers['Authorization'] = `Bearer ${accessToken.data.access_token}`;
            return Promise.resolve(config);
        },
        (err) => {
            return Promise.reject(err);
        });

    const makeRequest = async () => {
        dispatch(executing());
        let response;
        try {
            response = await instance[verb](`${path}`, params);
            dispatch(success(response));
        } catch (err) {
            if (axios.isCancel(err)) {
                console.log(err.message);
            } else {
                dispatch(error(err.response, err));
            }
        }
    };


    return [state, makeRequest];
};

const useApiRequest = (path, {verb = 'get', params = {}} = {}) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const instance = axios.create();
    const {envContext} = useEnvContext();
    const [keycloak] = useKeycloak();

    instance.interceptors.response.use(response => {
        return response;
    }, async error => {
        return handleError(instance, error, keycloak);
    });
    instance.interceptors.request.use(async (config) => configureAxios(envContext, config, keycloak),
        (err) => {
            return Promise.reject(err);
        });

    const makeRequest = async () => {
        dispatch(executing());
        let response;
        try {
            response = await instance[verb](`${envContext.url}${path}`, params);
            dispatch(success(response));
        } catch (err) {
            if (axios.isCancel(err)) {
                console.log(err.message);
            } else {
                dispatch(error(err.response, err));
            }
        }
    };


    return [state, makeRequest];
};

const handleError = async (instance, error, keycloak) => {
    const keycloakConfig = appConfig.get('keycloak');
    if (error.response && (error.response.status === 403 || error.response.status === 401)) {
        console.log("Retrying..." + error.response.status);
        let response;
        try {
            response = await axios({
                method: 'POST',
                url: `${keycloakConfig.authUrl}/realms/${keycloakConfig.realm}/protocol/openid-connect/token`,
                headers: {
                    "Content-Type": "application/x-www-form-urlencoded"
                },
                data: qs.stringify({
                    grant_type: 'refresh_token',
                    client_id: keycloakConfig.clientId,
                    refresh_token: keycloak.refreshToken
                })
            });
        } catch (e) {
            console.error(e);
        }
        console.log("Got new token");
        const token = response.data.access_token;
        const config = error.config;
        config.headers['Authorization'] = `Bearer ${token}`;
        return new Promise((resolve, reject) => {
            instance.request(config).then(response => {
                resolve(response);
            }).catch((err) => {
                reject(err);
            });
        });
    } else {
        return Promise.reject(error);
    }
};

export default useApiRequest;
