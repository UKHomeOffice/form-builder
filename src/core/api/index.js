import {useReducer} from 'react';
import axios from 'axios';
import reducer, {initialState} from './reducer';
import {error, executing, success} from './actionCreators';
import useEnvContext from "../context/useEnvContext";
import useLogger from "../logging/useLogger";
import {KeycloakTokenProvider} from "../KeycloakTokenProvider";
import secureLS from "../storage";
import {useKeycloak} from "react-keycloak";
import axiosRetry from 'axios-retry';

const keycloakTokenProvider = new KeycloakTokenProvider();

const configureAxios = async (envContext, config, keycloak) => {
    if (envContext) {
        const token = secureLS.get("jwt-token");
        config.headers['Accept'] = 'application/json';
        config.headers['Content-Type'] = 'application/json';
        config.headers['Cache-Control'] = "no-cache";
        config.headers['x-user-email'] = keycloak.tokenParsed.email;
        if (config.headers['x-promote-kc-token']) {
            config.headers.Authorization = config.headers['x-promote-kc-token'];
            delete config.headers['x-promote-kc-token'];
        } else {
            const jwtToken = await keycloakTokenProvider.fetchKeycloakToken(envContext, token);
            config.headers['Authorization'] = `Bearer ${jwtToken}`;
        }
    }
    return Promise.resolve(config);
};


const useApiRequest = (path, {verb = 'get', params = {}} = {}) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const instance = axios.create();
    const {envContext} = useEnvContext();
    const [keycloak] = useKeycloak();

    axiosRetry(instance, { retries: 3 });

    instance.interceptors.request.use(async (config) => configureAxios(envContext, config, keycloak),
        (err) => {
            return Promise.reject(err);
        });

    const makeRequest = async () => {
        dispatch(executing());
        let response;
        try {
            response = await instance[verb](`${envContext.url}${path}`, params);
            dispatch(success(response));
        } catch (err) {
            console.log(err.stack);
            if (axios.isCancel(err)) {
                console.log(err.message);
            } else {
                dispatch(error(err.response, err));
            }
        }
    };

    return [state, makeRequest];
};


export const useMultipleApiCallbackRequest = (apiCallback, logBefore = null, logAfter = null) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const instance = axios.create();
    const {envContext} = useEnvContext();
    const {log} = useLogger();
    const [keycloak] = useKeycloak();

    axiosRetry(instance, { retries: 3 });

    instance.interceptors.request.use(async (config) =>
            configureAxios(envContext, config, keycloak),
        (err) => {
            return Promise.reject(err);
        });

    const execute = async () => {
        if (logBefore) {
            log(logBefore);
        }
        dispatch(executing());
        try {
            const response = await apiCallback(instance);
            if (logAfter) {
                log(logAfter);
            }
            dispatch(success(response));
        } catch ({response = null, ...exception}) {
            log([{
                message: "An error occurred",
                exception,
                response,
                level: 'error'
            }]);
            dispatch(error(response, exception));
        }
    };
    return [state, execute];
};


export default useApiRequest;
