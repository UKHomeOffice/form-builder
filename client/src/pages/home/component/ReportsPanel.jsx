import {useTranslation} from "react-i18next";
import React from "react";
import useReports from "../useReports";
import {ResponsivePie} from '@nivo/pie'
import {ResponsiveBar} from "@nivo/bar";
import './ReportsPanel.scss';
import Row from "react-bootstrap/Row";
import Container from "react-bootstrap/Container";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import {EXECUTING} from "../../../core/api/actionTypes";
import Spinner from "react-bootstrap/Spinner";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faChartLine} from "@fortawesome/free-solid-svg-icons";

const ReportsPanel = () => {
    const {reports} = useReports();
    const {t} = useTranslation();
    return <Container fluid>
        <div style={{textAlign: 'center'}}>
            <h1 className="display-5"><FontAwesomeIcon icon={faChartLine}/> <span className="ml-2">{t('home.reports')}</span></h1>
        </div>

        <Row>
            <Col>
                <Card style={{width: '100hv'}}>
                    <Card.Body>
                        <Card.Title>
                            {t('home.forms-per-environment')}
                        </Card.Title>
                        {
                            reports.statusFormsPerEnvCount === EXECUTING ?
                                <div style={{textAlign: 'center'}}><Spinner animation="border" role="status"
                                                                            variant='success'>
                                    <span className="sr-only">{t('loading')}</span>
                                </Spinner></div>
                                : <div style={{height: '500px'}}><ResponsivePie
                                    data={reports.formsPerEnvCount}
                                    margin={{top: 40, right: 80, bottom: 80, left: 80}}
                                    innerRadius={0.5}
                                    padAngle={0.7}
                                    cornerRadius={3}
                                    colors={{scheme: 'nivo'}}
                                    borderWidth={1}
                                    borderColor={{from: 'color', modifiers: [['darker', 0.2]]}}
                                    radialLabelsSkipAngle={10}
                                    radialLabelsTextXOffset={6}
                                    radialLabelsTextColor="#333333"
                                    radialLabelsLinkOffset={0}
                                    radialLabelsLinkDiagonalLength={16}
                                    radialLabelsLinkHorizontalLength={24}
                                    radialLabelsLinkStrokeWidth={1}
                                    radialLabelsLinkColor={{from: 'color'}}
                                    slicesLabelsSkipAngle={10}
                                    slicesLabelsTextColor="#333333"
                                    animate={true}
                                    motionStiffness={90}
                                    motionDamping={15}
                                    legends={[
                                        {
                                            anchor: 'bottom',
                                            direction: 'row',
                                            translateY: 56,
                                            itemWidth: 100,
                                            itemHeight: 18,
                                            itemTextColor: '#999',
                                            symbolSize: 18,
                                            symbolShape: 'circle',
                                            effects: [
                                                {
                                                    on: 'hover',
                                                    style: {
                                                        itemTextColor: '#000'
                                                    }
                                                }
                                            ]
                                        }
                                    ]}
                                /></div>
                        }

                    </Card.Body>
                </Card>

            </Col>
            <Col>
                <Card style={{width: '100hv'}}>
                    <Card.Body>
                        <Card.Title>
                            {t('home.types-of-forms')}
                        </Card.Title>

                        {
                            reports.statusTypeData === EXECUTING ?
                                <div style={{textAlign: 'center'}}><Spinner animation="border" role="status"
                                                                            variant='success'>
                                    <span className="sr-only">{t('loading')}</span>
                                </Spinner></div> : <div style={{height: '500px'}}><ResponsiveBar
                                    data={reports.typeData}
                                    keys={['wizard', 'form']}
                                    indexBy="name"
                                    margin={{top: 50, right: 130, bottom: 50, left: 60}}
                                    padding={0.3}
                                    colors={{scheme: 'nivo'}}
                                    axisTop={null}
                                    axisRight={null}
                                    axisBottom={{
                                        tickSize: 5,
                                        tickPadding: 5,
                                        tickRotation: 0,
                                        legend: 'Environments',
                                        legendPosition: 'middle',
                                        legendOffset: 32
                                    }}
                                    axisLeft={{
                                        tickSize: 5,
                                        tickPadding: 5,
                                        tickRotation: 0,
                                        legend: 'Total',
                                        legendPosition: 'middle',
                                        legendOffset: -40
                                    }}
                                    labelSkipWidth={12}
                                    labelSkipHeight={12}
                                    labelTextColor={{from: 'color', modifiers: [['darker', 1.6]]}}
                                    legends={[
                                        {
                                            dataFrom: 'keys',
                                            anchor: 'bottom-right',
                                            direction: 'column',
                                            justify: false,
                                            translateX: 120,
                                            translateY: 0,
                                            itemsSpacing: 2,
                                            itemWidth: 100,
                                            itemHeight: 20,
                                            itemDirection: 'left-to-right',
                                            itemOpacity: 0.85,
                                            symbolSize: 20,
                                            effects: [
                                                {
                                                    on: 'hover',
                                                    style: {
                                                        itemOpacity: 1
                                                    }
                                                }
                                            ]
                                        }
                                    ]}

                                    animate={true}
                                    motionStiffness={90}
                                    motionDamping={15}
                                /></div>
                        }

                    </Card.Body>
                </Card>

            </Col>
        </Row>
    </Container>
};

export default ReportsPanel;
