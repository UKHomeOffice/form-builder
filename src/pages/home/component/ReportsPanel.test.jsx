import {SUCCESS} from "../../../core/api/actionTypes";
import ReportsPanel from "./ReportsPanel";
import React from "react";
import {mount} from "enzyme";

describe("ReportsPanel", () => {
    beforeEach(() => {
        const useReports = require('../useReports');
        useReports.default = () => {

            const reports = {
                statusFormsPerEnvCount: SUCCESS,
                activeIndex: 0,
                formsPerEnvCount: [{
                    "id": "dev",
                    "label": "Development",
                    "value": 10
                }, {
                    "id": "demo",
                    "label": "Demo",
                    "value": 12
                }],
                typeData: [{
                    "name": "dev",
                    "wizard": 10,
                    "form": 12
                }, {
                    "name": "demo",
                    "wizard": 10,
                    "form": 12
                }],
                statusTypeData: SUCCESS
            };
            return {
                reports: reports
            }
        }
    });

    it('renders pie and bar chart', async () => {
        const wrapper = mount(<ReportsPanel/>);
        expect(wrapper).toMatchSnapshot();
    });
});
