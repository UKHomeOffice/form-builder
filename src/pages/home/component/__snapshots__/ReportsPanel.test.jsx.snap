// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsPanel renders pie and bar chart 1`] = `
<ReportsPanel>
  <Grid
    stackable={true}
  >
    <div
      className="ui stackable grid"
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            width={8}
          >
            <div
              className="eight wide column"
            >
              <Segment
                className="chart-container"
                loading={false}
              >
                <div
                  className="ui segment chart-container"
                >
                  <Header
                    as="h3"
                    dividing={true}
                  >
                    <h3
                      className="ui dividing header"
                    >
                      home.forms-per-environment
                    </h3>
                  </Header>
                  <div
                    style={
                      Object {
                        "height": "500px",
                      }
                    }
                  >
                    <ResponsivePie
                      animate={true}
                      borderColor={
                        Object {
                          "from": "color",
                          "modifiers": Array [
                            Array [
                              "darker",
                              0.2,
                            ],
                          ],
                        }
                      }
                      borderWidth={1}
                      colors={
                        Object {
                          "scheme": "nivo",
                        }
                      }
                      cornerRadius={3}
                      data={
                        Array [
                          Object {
                            "id": "dev",
                            "label": "Development",
                            "value": 10,
                          },
                          Object {
                            "id": "demo",
                            "label": "Demo",
                            "value": 12,
                          },
                        ]
                      }
                      innerRadius={0.5}
                      legends={
                        Array [
                          Object {
                            "anchor": "bottom",
                            "direction": "row",
                            "effects": Array [
                              Object {
                                "on": "hover",
                                "style": Object {
                                  "itemTextColor": "#000",
                                },
                              },
                            ],
                            "itemHeight": 18,
                            "itemTextColor": "#999",
                            "itemWidth": 100,
                            "symbolShape": "circle",
                            "symbolSize": 18,
                            "translateY": 56,
                          },
                        ]
                      }
                      margin={
                        Object {
                          "bottom": 80,
                          "left": 80,
                          "right": 80,
                          "top": 40,
                        }
                      }
                      motionDamping={15}
                      motionStiffness={90}
                      padAngle={0.7}
                      radialLabelsLinkColor={
                        Object {
                          "from": "color",
                        }
                      }
                      radialLabelsLinkDiagonalLength={16}
                      radialLabelsLinkHorizontalLength={24}
                      radialLabelsLinkOffset={0}
                      radialLabelsLinkStrokeWidth={1}
                      radialLabelsSkipAngle={10}
                      radialLabelsTextColor="#333333"
                      radialLabelsTextXOffset={6}
                      slicesLabelsSkipAngle={10}
                      slicesLabelsTextColor="#333333"
                    >
                      <ResponsiveWrapper>
                        <Measure
                          bounds={true}
                          onResize={[Function]}
                        >
                          <Component
                            bounds={true}
                            contentRect={
                              Object {
                                "bounds": Object {},
                                "client": Object {},
                                "entry": Object {},
                                "margin": Object {},
                                "offset": Object {},
                                "scroll": Object {},
                              }
                            }
                            measure={[Function]}
                            measureRef={[Function]}
                          >
                            <div
                              style={
                                Object {
                                  "height": "100%",
                                  "width": "100%",
                                }
                              }
                            />
                          </Component>
                        </Measure>
                      </ResponsiveWrapper>
                    </ResponsivePie>
                  </div>
                </div>
              </Segment>
            </div>
          </GridColumn>
          <GridColumn
            width={8}
          >
            <div
              className="eight wide column"
            >
              <Segment
                className="chart-container"
                loading={false}
              >
                <div
                  className="ui segment chart-container"
                >
                  <Header
                    as="h3"
                    dividing={true}
                  >
                    <h3
                      className="ui dividing header"
                    >
                      home.types-of-forms
                    </h3>
                  </Header>
                  <div
                    style={
                      Object {
                        "height": "500px",
                      }
                    }
                  >
                    <ResponsiveBar
                      animate={true}
                      axisBottom={
                        Object {
                          "legend": "Environments",
                          "legendOffset": 32,
                          "legendPosition": "middle",
                          "tickPadding": 5,
                          "tickRotation": 0,
                          "tickSize": 5,
                        }
                      }
                      axisLeft={
                        Object {
                          "legend": "Total",
                          "legendOffset": -40,
                          "legendPosition": "middle",
                          "tickPadding": 5,
                          "tickRotation": 0,
                          "tickSize": 5,
                        }
                      }
                      axisRight={null}
                      axisTop={null}
                      colors={
                        Object {
                          "scheme": "nivo",
                        }
                      }
                      data={
                        Array [
                          Object {
                            "form": 12,
                            "name": "dev",
                            "wizard": 10,
                          },
                          Object {
                            "form": 12,
                            "name": "demo",
                            "wizard": 10,
                          },
                        ]
                      }
                      indexBy="name"
                      keys={
                        Array [
                          "wizard",
                          "form",
                        ]
                      }
                      labelSkipHeight={12}
                      labelSkipWidth={12}
                      labelTextColor={
                        Object {
                          "from": "color",
                          "modifiers": Array [
                            Array [
                              "darker",
                              1.6,
                            ],
                          ],
                        }
                      }
                      legends={
                        Array [
                          Object {
                            "anchor": "bottom-right",
                            "dataFrom": "keys",
                            "direction": "column",
                            "effects": Array [
                              Object {
                                "on": "hover",
                                "style": Object {
                                  "itemOpacity": 1,
                                },
                              },
                            ],
                            "itemDirection": "left-to-right",
                            "itemHeight": 20,
                            "itemOpacity": 0.85,
                            "itemWidth": 100,
                            "itemsSpacing": 2,
                            "justify": false,
                            "symbolSize": 20,
                            "translateX": 120,
                            "translateY": 0,
                          },
                        ]
                      }
                      margin={
                        Object {
                          "bottom": 50,
                          "left": 60,
                          "right": 130,
                          "top": 50,
                        }
                      }
                      motionDamping={15}
                      motionStiffness={90}
                      padding={0.3}
                    >
                      <ResponsiveWrapper>
                        <Measure
                          bounds={true}
                          onResize={[Function]}
                        >
                          <Component
                            bounds={true}
                            contentRect={
                              Object {
                                "bounds": Object {},
                                "client": Object {},
                                "entry": Object {},
                                "margin": Object {},
                                "offset": Object {},
                                "scroll": Object {},
                              }
                            }
                            measure={[Function]}
                            measureRef={[Function]}
                          >
                            <div
                              style={
                                Object {
                                  "height": "100%",
                                  "width": "100%",
                                }
                              }
                            />
                          </Component>
                        </Measure>
                      </ResponsiveWrapper>
                    </ResponsiveBar>
                  </div>
                </div>
              </Segment>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </Grid>
</ReportsPanel>
`;
